pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BookkeeperBucketSnapshotStorage.java

```diff
@@ -0,0 +1,246 @@
+        } catch (IOException e) {
+            result.completeExceptionally(e);
+        }
+        return result;
+    }
+
+    @NotNull
+    private CompletableFuture<LedgerHandle> createLedger() {
+        CompletableFuture<LedgerHandle> future = new CompletableFuture<>();
+        bookKeeper.asyncCreateLedger(
+                config.getManagedLedgerDefaultEnsembleSize(),
+                config.getManagedLedgerDefaultWriteQuorum(),
+                config.getManagedLedgerDefaultAckQuorum(),
+                BookKeeper.DigestType.fromApiDigestType(config.getManagedLedgerDigestType()),
+                LedgerPassword,
+                (rc, handle, ctx) -> {
+                    if (rc != BKException.Code.OK) {
+                        future.completeExceptionally(bkException("Failed to create ledger", rc, -1));
+                    } else {
+                        future.complete(handle);
+                    }
+                }, null, null);
+        return future;
+    }
+
+    private CompletableFuture<LedgerHandle> openLedger(Long ledgerId) {
+        final CompletableFuture<LedgerHandle> future = new CompletableFuture<>();
+        bookKeeper.asyncOpenLedger(
+                ledgerId,
+                BookKeeper.DigestType.fromApiDigestType(config.getManagedLedgerDigestType()),
+                LedgerPassword,
+                (rc, handle, ctx) -> {
+                    if (rc != BKException.Code.OK) {
+                        future.completeExceptionally(bkException("Failed to open ledger", rc, ledgerId));
+                    } else {
+                        future.complete(handle);
+                    }
+                }, null
+        );
+        return future;
+    }
+
+    private CompletableFuture<Void> closeLedger(LedgerHandle ledgerHandle) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        ledgerHandle.asyncClose((rc, handle, ctx) -> {
+            if (rc != BKException.Code.OK) {
+                log.warn("Failed to close a Ledger Handle: {}", ledgerHandle.getId());
+                future.completeExceptionally(bkException("Failed to close ledger", rc, ledgerHandle.getId()));
+            } else {
+                future.complete(null);
+            }
+        }, null);
+        return future;
+    }
+
+    private CompletableFuture<Void> addEntry(LedgerHandle ledgerHandle, byte[] data) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        ledgerHandle.asyncAddEntry(data,
+                (rc, handle, entryId, ctx) -> {
+                    if (rc != BKException.Code.OK) {
+                        future.completeExceptionally(bkException("Failed to add entry", rc, ledgerHandle.getId()));
+                    } else {
+                        future.complete(null);
+                    }
+                }, null
+        );
+        return future;
+    }
+
+    CompletableFuture<Enumeration<LedgerEntry>> getLedgerEntry(LedgerHandle ledger,
```

Comment:
`addEntry` will not close the ledger, but `getLedgerEntry` will close the ledger, suggestion change this method name to `getLedgerEntryThenCloseLedger`.

Reply:
Ok.